packages/router-core/src/routeInfo.ts

```diff
@@ -0,0 +1,24 @@
+export type ParseRoute<TRouteTree, TAcc = TRouteTree> = TRouteTree extends {
+  types: { children: infer TChildren }
+}
+  ? unknown extends TChildren
+    ? TAcc
+    : TChildren extends ReadonlyArray<any>
+      ? ParseRoute<TChildren[number], TAcc | TChildren[number]>
+      : ParseRoute<
+          TChildren[keyof TChildren],
+          TAcc | TChildren[keyof TChildren]
+        >
+  : TAcc
+
+export type ParentPath<TOption> = 'always' extends TOption
```

Comment:
These functions can't be extracted this easily - they depend on the AnyRouter now:

https://github.com/TanStack/router/blob/255f31603d0aed10bfdfd25bf403524c6238c5ac/packages/react-router/src/routeInfo.ts#L62-L75